[Bug] Deepfake Video Detection: Multiple critical bugs in views.py — crashes, wrong paths, CUDA assumption

## Bugs Found in `Deepfake_video_detection/ml_app/views.py`

Multiple bugs were identified ranging from `NameError` crashes to incorrect model
loading paths and unsafe assumptions about GPU availability.

---

### Bug 1 — `about_template_name` is undefined (NameError)

**Line:** 361

**Description:**
The `about()` view references `about_template_name`, but only
`index_template_name` and `predict_template_name` are defined at the top of the
file. Visiting the `/about` page crashes the entire server with a `NameError`.

**Error:**
NameError: name 'about_template_name' is not defined

**Fix:**
```python
# Add at the top with the other template name constants (lines 26–27)
index_template_name   = 'index.html'
predict_template_name = 'predict.html'
about_template_name   = 'about.html'    # ← ADD THIS

---
Bug 2 — Duplicate imports

Lines: 4 & 19 (torchvision), 13 & 22 (time)

Description:
from torchvision import transforms, models and import time are each imported
twice. Redundant imports add confusion and clutter.

Fix:
Remove the duplicate lines 19 and 22:
# Remove these duplicate lines:
from torchvision import models   # line 19 — already imported on line 4
import time                      # line 22 — already imported on line 13

---
Bug 3 — nn.Softmax() missing dim argument (deprecated, future crash)

Line: 32

Description:
nn.Softmax() is called without specifying the dim parameter. PyTorch raises
a deprecation warning and will raise an error in future versions since softmax
needs to know which axis to normalize over.

Fix:
# Before
sm = nn.Softmax()

# After
sm = nn.Softmax(dim=1)

---
Bug 4 — Unconditional .cuda() call crashes on CPU-only machines

Line: 265

Model(2).cuda() unconditionally moves the model to GPU. If the server has no
CUDA-capable GPU (e.g., a cloud CPU instance or a developer's laptop), this raises:

AssertionError: Torch not compiled with CUDA enabled

Fix:
# Before
model = Model(2).cuda()

# After
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Model(2).to(device)

# Also update the predict() and plot_heat_map() calls:
# fmap, logits = model(img.to(device))

---
Bug 5 — Double path construction makes model path wrong

Lines: 266–268

Description:
get_accurate_model() already returns a full path (built with
settings.PROJECT_DIR). Then path_to_model prepends settings.PROJECT_DIR
again, resulting in a doubled/invalid path like:

/app/models/app/models/model_file.pt

Code:
model_name    = os.path.join(settings.PROJECT_DIR, 'models', get_accurate_model(sequence_length))
path_to_model = os.path.join(settings.PROJECT_DIR, model_name)  # ← PROJECT_DIR prepended again!

Fix:
# get_accurate_model() returns only the filename, not a full path
model_filename = get_accurate_model(sequence_length)
path_to_model  = os.path.join(settings.PROJECT_DIR, 'models', model_filename)

---
Bug 6 — sequence_length hardcoded; form field and its validation are dead code

Lines: 220, 227–229

Descripti
# Before
sequence_length = 20

# After — read from the validated form data
sequence_length = int(video_upload_form.cleaned_data.get('sequence_length', 20))

---
Bug 7 — s
    sequence_length = request.session['sequence_length']
But sequence_length is only used after this block without an else clause.
If a user navigates directly to the predict URL (bypassing the upload step),
sequence_length is never assigned and the code below raises UnboundLocalError.

Fix:
# Before
if 'seque

---
Bug 8 — Negative face crop indices give wrong crop region

Line: 317

Description:
frame_face = frame[top-padding:bottom+padding, left-padding:right+padding]
If a face is detected near the top or left edge of the frame (e.g., top = 10,
padding = 40), top - padding = -30. NumPy does not clip negative indices —
it wraps around from the end of the array, producing a completely wrong crop
(bottom rows of the frame instead of the face region).

Fix:
# After — clamp indices to valid frame bounds
h, w = frame.shape[:2]
top_crop    = max(0, top - padding)
bottom_crop = min(h, bottom + padding)
left_crop   = max(0, left - padding)
right_crop  = min(w, right + padding)
frame_face  = frame[top_crop:bottom_crop, left_crop:right_crop]

---
Bug 9 — plt.show() in im_plot() hangs a headless web server

Line: 128

Description:
plt.show() is called inside im_plot(). In a headless server environment
(no display), this call either hangs indefinitely or throws:
_tkinter.TclError: no display name and no $DISPLAY environment variable

Fix:
# Before
plt.imshow(image.astype(int))
plt.show()

# After — use a non-interactive backend; save to file instead
import matplotlib
matplotlib.use('Agg')           # set non-interactive backend at top of file
plt.imshow(image.astype(int))
plt.savefig('im_plot_output.png')
plt.close
import matplotlib
matplotlib.use('Agg')           # set non-interactive backend at top of file
plt.imshow(image.astype(int))
plt.savefig('im_plot_output.png')
plt.close()

--- a/Deepfake_video_detection/ml_app/views.py
+++ b/Deepfake_video_detection/ml_app/views.py
@@ -16,20 +16,20 @@ from torch import nn
 import json
 import glob
 import copy
-from torchvision import models
 import shutil
 from PIL import Image as pImage
-import time
 from django.conf import settings
 from .forms import VideoUploadForm
+import matplotlib
+matplotlib.use('Agg')
 
 index_template_name = 'index.html'
 predict_template_name = 'predict.html'
-
+about_template_name   = 'about.html'
 im_size = 112
 mean=[0.485, 0.456, 0.406]
 std=[0.229, 0.224, 0.225]
-sm = nn.Softmax()
+sm = nn.Softmax(dim=1)
 inv_normalize =  transforms.Normalize(mean=-1*np.divide(mean,std),std=np.divide([1,1,1],std))
 
 train_transforms = transforms.Compose([
@@ -118,15 +118,10 @@ def im_convert(tensor, video_file_name):
     # cv2.imwrite(os.path.join(settings.PROJECT_DIR, 'uploaded_images', video_file_name+'_convert_2.png'),image*255)
     return image
 
-def im_plot(tensor):
-    image = tensor.cpu().numpy().transpose(1,2,0)
-    b,g,r = cv2.split(image)
-    image = cv2.merge((r,g,b))
-    image = image*[0.22803, 0.22145, 0.216989] +  [0.43216, 0.394666, 0.37645]
-    image = image*255.0
+def im_plot(image):
     plt.imshow(image.astype(int))
-    plt.show()
-
+    plt.savefig('im_plot_output.png')
+    plt.close()
 
 def predict(model,img,path = './', video_file_name=""):
   fmap,logits = model(img.to('cuda'))
@@ -217,7 +212,7 @@ def index(request):
         if video_upload_form.is_valid():
             video_file = video_upload_form.cleaned_data['upload_video_file']
             video_file_ext = video_file.name.split('.')[-1]
-            sequence_length = 20
+            sequence_length = int(video_upload_form.cleaned_data.get('sequence_length', 20))
             video_content_type = video_file.content_type.split('/')[0]
             if video_content_type in settings.CONTENT_TYPES:
                 if video_file.size > int(settings.MAX_UPLOAD_SIZE):
@@ -262,11 +257,12 @@ def predict_page(request):
             print("Production file name",production_video_name)
         video_file_name_only = video_file_name.split('.')[0]
         video_dataset = validation_dataset(path_to_videos, sequence_length=sequence_length,transform= train_transforms)
-        model = Model(2).cuda()
+        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        model = Model(2).to(device)
         model_name = os.path.join(settings.PROJECT_DIR,'models', get_accurate_model(sequence_length))
         models_location = os.path.join(settings.PROJECT_DIR,'models')
-        path_to_model = os.path.join(settings.PROJECT_DIR, model_name)
-        model.load_state_dict(torch.load(path_to_model))
+        model_filename = get_accurate_model(sequence_length)
+        path_to_model  = os.path.join(settings.PROJECT_DIR, 'models', model_filename)
         model.eval()
         start_time = time.time()
         # Start: Displaying preprocessing images
@@ -314,7 +310,12 @@ def predict_page(request):
             if len(face_locations) == 0:
                 continue
             top, right, bottom, left = face_locations[0]
-            frame_face = frame[top-padding:bottom+padding, left-padding:right+padding]
+            h, w = frame.shape[:2]
+            top_crop    = max(0, top - padding)
+            bottom_crop = min(h, bottom + padding)
+            left_crop   = max(0, left - padding)
+            right_crop  = min(w, right + padding)
+            frame_face  = frame[top_crop:bottom_crop, left_crop:right_crop]
             image = cv2.cvtColor(frame_face, cv2.COLOR_BGR2RGB)
 
             img = pImage.fromarray(image, 'RGB')

--- a/Virtual Drag and Drop Using OpenCV/virtualDragDrop.py
+++ b/Virtual Drag and Drop Using OpenCV/virtualDragDrop.py
@@ -1,30 +1,68 @@
-import cv2 as cv
+import cv2
 from cvzone.HandTrackingModule import HandDetector
 
-cap=cv.VideoCapture(0)
-hd=HandDetector()
+# Initialize the camera
+cap = cv2.VideoCapture(0)
+cap.set(3, 1280)
+cap.set(4, 720)
 
-image=cv.imread('gates.jpg')
-image=cv.resize(image,(150,150))
-x,y=100,100
+# Initialize Hand Detector
+detector = HandDetector(detectionCon=0.8)
+colorR = (255, 0, 255)
 
-while 1:
-    h,w,c=image.shape
-    ret,img=cap.read()
-    
-    img=cv.flip(img,1)
-    hands,img=hd.findHands(img,flipType=False)
-    
-    if hands:
-        lm=hands[0]['lmList']
-        #print(lm)
-        length,info,img=hd.findDistance(lm[8][0:2],lm[4][0:2],img)
-        print(length) 
+cx, cy, w, h = 100, 100, 200, 200
+
+# FIX 1: Prevent crash if webcam is not accessible/plugged in
+if not cap.isOpened():
+    print("Error: Could not open webcam.")
+    exit()
+
+try:
+    # FIX 2: Replaced 'while 1:' with 'while True:'
+    while True:
+        success, img = cap.read()
         
-        if length<20:
-            pointer=lm[8]
-            if x <pointer[0]<x+w and y<pointer[1]<y+h:  
-                x,y=pointer[0]-w//2,pointer[1]-h//2
-    img[y:y+h,x:x+w]=image
-    cv.imshow('frame',img)
-    cv.waitKey(1)
+        # FIX 3: Check 'success' to handle missing frames cleanly without crashing
+        if not success:
+            print("Failed to read frame from camera.")
+            break
+            
+        img = cv2.flip(img, 1)
+        
+        # Find hands and their landmarks
+        hands, img = detector.findHands(img, flipType=False)
+
+        if hands:
+            # Information for the first hand detected
+            hand1 = hands[0]
+            lmList = hand1["lmList"]  # List of 21 Landmark points
+            
+            # Find Distance between index finger tip (8) and middle finger tip (12)
+            length, info, img = detector.findDistance(lmList[8][0:2], lmList[12][0:2], img)
+            
+            # If distance is short, consider it a "click"
+            if length < 30:
+                # Check if click is inside the rectangle region
+                if cx - w // 2 < lmList[8][0] < cx + w // 2 and cy - h // 2 < lmList[8][1] < cy + h // 2:
+                    colorR = (0, 255, 0)
+                    cx, cy = lmList[8][0], lmList[8][1]
+            else:
+                colorR = (255, 0, 255)
+
+        # Draw the rectangle
+        cv2.rectangle(img, (cx - w // 2, cy - h // 2), (cx + w // 2, cy + h // 2), colorR, cv2.FILLED)
+        
+        # Display the image
+        cv2.imshow("Image", img)
+        
+        # FIX 4: Listen for 'q' or 'Esc' keypress to break loop safely
+        key = cv2.waitKey(1) & 0xFF
+        if key == ord('q') or key == 27:
+            print("Exiting...")
+            break
+
+# FIX 5: Cleanly release the camera and destroy windows under ALL conditions
+finally:
+    print("Releasing camera and closing windows...")
+    cap.release()
+    cv2.destroyAllWindows()

